### PR TITLE
page_iterator_sync is added. [API-1761]

### PR DESCRIPTION
**Pull request by OzanCansel**
- `page_iterator_sync` is added.
- `pbegin(std::chrono::milliseconds)` and `pend()` methods are added.
- `sql_page_iterator_sync` example is added.

`page_iterator_sync` simply wraps `page_iterator` and behaves as syncronous iterator. It allows to define timeout for operations(it is required to prevent from infinite blocking).

### CODE REVIEW

File: Reference_Manual.md [hunk 1]
```diff
@@ -3683,10 +3686,59 @@ Exceptions which can be thrown by SQL API are stated at below:
 - `sql_result::row_metadata()` can throw `illegal_state` exception if the result contains only update count.
 - `sql_page::sql_row::get_object(int)` can throw `index_out_of_bounds` exception if the index is out of range.
 - `sql_page::sql_row::get_object(std::string)` can throw `illegal_argument` exception if the column doesn't exist.
+- `sql_result::page_iterator_sync::operator++()` can throw `no_such_element` exception if the fetch operation is timedout.
```
Comment (akeles85) [inline Reference_Manual.md:3689]: timed out

Comment (OzanCansel) [reply Reference_Manual.md:3689]: Fixed.

File: Reference_Manual.md [hunk 2]
```diff
@@ -3683,10 +3686,59 @@ Exceptions which can be thrown by SQL API are stated at below:
...
 
+### 7.11.8 Iterators
+There are several iterators in SQL Api to satisfy different use cases. They are listed below:
+
+- `page_iterator` (Async page iterator)
+- `page_iterator_sync` (Sync page iterator)
+
+Reminder 1, only one type of iterator can be used. So it is suggested to choose appropriate one to fetch `SELECT` query results.
+
+Reminder 2, it is not possible iterate from beginning to end more than once.
```
[the remark this answers is not in the document]
Comment (OzanCansel) [reply Reference_Manual.md:3703]: Fixed.

File: Reference_Manual.md [hunk 3]
```diff
@@ -3683,10 +3686,59 @@ Exceptions which can be thrown by SQL API are stated at below:
...
+        std::cout << "(" << row.get_object<int>(0) << ")" << std::endl;
+    }
+}
+```
+
+#### 7.11.8.2 page_iterator_sync
+`page_iterator_sync` is the second iterator which wraps `page_iterator` and serves it as a sync iterator. It allows users to iterator over pages.
+It is similar to native C++ iterators. So it can be used with `algorithm` header.
+It supports `operator*()`, `operator->()`, `operator++()` operators. Post increment operator is marked as `delete`.
+`operator++()`(pre-increment operator) fetches the next page in a blocking manner. `timeout` can be set. `operator++()` might throw `exception::no_such_element` exception if the fetch operation is timedout.
```
[the remark this answers is not in the document]
Comment (OzanCansel) [reply Reference_Manual.md:3730]: Fixed.

File: hazelcast/test/src/sql_test.cpp [hunk 4]
```diff
@@ -298,21 +309,6 @@ class SqlTest : public ClientTest
...
     }
 
-    static void TearDownTestSuite()
-    {
-        member_.reset();
-        member2_.reset();
-        server_factory_.reset();
-    }
-
-    void SetUp() override
```
Comment (akeles85) [inline hazelcast/test/src/sql_test.cpp:308]: Why did you delete this method?

File: hazelcast/include/hazelcast/client/sql/sql_result.h [hunk 5]
```diff
@@ -100,6 +107,77 @@ class HAZELCAST_API sql_result : public std::enable_shared_from_this<sql_result>
...
+    class HAZELCAST_API page_iterator_sync
+    {
+    public:
+        using difference_type = void;
+        using value_type = std::shared_ptr<sql_page>;
+        using pointer = std::shared_ptr<sql_page>;
+        using reference = std::shared_ptr<sql_page>&;
+        using iterator_category = std::input_iterator_tag;
+
+        page_iterator_sync(page_iterator&&, std::chrono::milliseconds timeout);
```
Comment (akeles85) [inline hazelcast/include/hazelcast/client/sql/sql_result.h:122]: It might throw exception because of check_closed method, could you document it?

File: hazelcast/include/hazelcast/client/sql/sql_result.h [hunk 6]
```diff
@@ -153,6 +231,10 @@ class HAZELCAST_API sql_result : public std::enable_shared_from_this<sql_result>
      */
     page_iterator iterator();
 
+    page_iterator_sync pbegin(
```
Comment (akeles85) [inline hazelcast/include/hazelcast/client/sql/sql_result.h:234]: It might throw exceptions like illegal_state, could you document it?

File: hazelcast/test/src/sql_test.cpp [hunk 7]
```diff
@@ -1463,6 +1506,57 @@ TEST_F(SqlTest, exception)
...
+    auto result =
+      client.get_sql().execute("SELECT * FROM TABLE(generate_stream(1))").get();
+
+    auto statement = [&result](){
+        auto it = result->pbegin(std::chrono::milliseconds{ 1 });
+        ++it;
+        ++it;
+    };
+
+    ASSERT_THROW(statement(), hazelcast::client::exception::no_such_element);
```
Comment (akeles85) [inline hazelcast/test/src/sql_test.cpp:1557]: I would prefer a different exception type like timeout as a user of library.

File: hazelcast/src/hazelcast/client/sql.cpp [hunk 8]
```diff
@@ -828,6 +840,100 @@ sql_result::page_iterator::has_next() const
...
 }
 
+sql_result::page_iterator_sync::non_copyables::non_copyables(
+  page_iterator&& iter)
+  : preloaded_page_{}
+  , iter_{ std::move(iter) }
+{
+}
+
+sql_result::page_iterator_sync::page_iterator_sync(
```
Comment (akeles85) [inline hazelcast/src/hazelcast/client/sql.cpp:850]: I suggest this method to be private and sql_result to be a friend of this class. It seems user does not need it.

File: hazelcast/src/hazelcast/client/sql.cpp [hunk 9]
```diff
@@ -828,6 +840,100 @@ sql_result::page_iterator::has_next() const
...
+  , current_{ block_->iter_.next().get() }
+  , timeout_{ timeout }
+{
+    if (block_->iter_.has_next()) {
+        block_->preloaded_page_ = block_->iter_.next();
+    }
+}
+
+void
+sql_result::page_iterator_sync::set_timeout(std::chrono::milliseconds t)
```
Comment (akeles85) [inline hazelcast/src/hazelcast/client/sql.cpp:863]: I suggest this method to be private and sql_result to be a friend of this class. It seems user does not need it.

File: hazelcast/src/hazelcast/client/sql.cpp [hunk 10]
```diff
@@ -828,6 +840,100 @@ sql_result::page_iterator::has_next() const
...
+{
+    timeout_ = t;
+}
+
+std::chrono::milliseconds
+sql_result::page_iterator_sync::timeout() const
```
Comment (akeles85) [inline hazelcast/src/hazelcast/client/sql.cpp:869]: I suggest this method to be private and sql_result to be a friend of this class. It seems user does not need it.

**Review by akeles85 — CHANGES_REQUESTED** [review]
I have some comments for requesting changes.

File: hazelcast/src/hazelcast/client/sql.cpp (hunk 10 above)
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/sql.cpp:869]: I designed it as public. User might want to retrieve the timeout which is `set` at somewhere else. What's harm in it ?

File: hazelcast/src/hazelcast/client/sql.cpp (hunk 8 above)
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/sql.cpp:850]: Fixed.

File: hazelcast/test/src/sql_test.cpp (hunk 7 above)
Comment (OzanCansel) [reply hazelcast/test/src/sql_test.cpp:1557]: By mistaken :)

Fixed.

File: hazelcast/include/hazelcast/client/sql/sql_result.h (hunk 6 above)
Comment (OzanCansel) [reply hazelcast/include/hazelcast/client/sql/sql_result.h:234]: Fixed.

File: hazelcast/include/hazelcast/client/sql/sql_result.h (hunk 5 above)
Comment (OzanCansel) [reply hazelcast/include/hazelcast/client/sql/sql_result.h:122]: Fixed.

File: hazelcast/test/src/sql_test.cpp (hunk 4 above)
Comment (OzanCansel) [reply hazelcast/test/src/sql_test.cpp:308]: By mistaken :)

Fixed.

File: hazelcast/src/hazelcast/client/sql.cpp (hunk 10 above)
Comment (akeles85) [reply hazelcast/src/hazelcast/client/sql.cpp:869]: Ok, no harm.
